demos/mqtt/mqtt_demo_connection_sharing/subscription-manager/mqtt_subscription_manager.c

```diff
@@ -0,0 +1,288 @@
+        }
+    }
+
+    *pMatch = status;
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static bool matchWildcards( const char * pTopicFilter,
+                            const char * pTopicName,
+                            uint16_t topicNameLength,
+                            uint16_t filterIndex,
+                            uint16_t * pNameIndex,
+                            bool * pMatch )
+{
+    bool status = false;
+
+    /* Check for wildcards. */
+    if( pTopicFilter[ filterIndex ] == '+' )
+    {
+        /* Move topic name index to the end of the current level.
+         * This is identified by '/'. */
+        while( ( *pNameIndex < topicNameLength ) && ( pTopicName[ *pNameIndex ] != '/' ) )
+        {
+            ( *pNameIndex )++;
+        }
+
+        ( *pNameIndex )--;
+    }
+    else if( pTopicFilter[ filterIndex ] == '#' )
+    {
+        /* Subsequent characters don't need to be checked for the
+         * multi-level wildcard. */
+        *pMatch = true;
+        status = true;
+    }
+    else
+    {
+        /* Any character mismatch other than '+' or '#' means the topic
+         * name does not match the topic filter. */
+        *pMatch = false;
+        status = true;
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static bool topicFilterMatch( const char * pTopicName,
+                              uint16_t topicNameLength,
+                              const char * pTopicFilter,
+                              uint16_t topicFilterLength )
+{
+    bool status = false, matchFound = false;
+    uint16_t nameIndex = 0, filterIndex = 0;
+
+    while( ( nameIndex < topicNameLength ) && ( filterIndex < topicFilterLength ) )
+    {
+        /* Check if the character in the topic name matches the corresponding
+         * character in the topic filter string. */
+        if( pTopicName[ nameIndex ] == pTopicFilter[ filterIndex ] )
+        {
+            /* Handle special corner cases regarding wildcards at the end of
+             * topic filters, as documented by the MQTT protocol spec. */
+            matchFound = matchEndWildcards( pTopicFilter,
+                                            topicNameLength,
+                                            topicFilterLength,
+                                            nameIndex,
+                                            filterIndex,
+                                            &status );
+        }
+        else
+        {
+            /* Check for matching wildcards. */
+            matchFound = matchWildcards( pTopicFilter,
+                                         pTopicName,
+                                         topicNameLength,
+                                         filterIndex,
+                                         &nameIndex,
+                                         &status );
+        }
```

Comment:
Don't we have to check for the topics beginning with $ as well?

Reply:
I think the '$' prefix reservation is specific to AWS IoT, so I'd think no

Reply:
http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718107, "Topics beginning with $", it is part of the spec. Not really specific to AWS IoT.

Reply:
The spec mentions that "_The Server MUST NOT match Topic Filters starting with a wildcard character (# or +) with Topic Names beginning with a $ character_".
The client does not need to perform any special handling for topics starting with "$". Our service (Jobs, Shadow, Provisioning, etc) subscribe to these topics (starting with "$") and the topic matching on receiving incoming PUBLISHes only involves performing an exact character match for the starting "$" in the topic 🙂

Reply:
how does the topic matching happens if there are 2 subscriptions in the client for topics “+/monitor/Clients” and “$SYS/monitor/Clients”?

Reply:
Good point, have added logic to avoid wildcard match for incoming PUBLISH topic names starting with the "$" character